### PR TITLE
Test observer waveform polarization.

### PR DESCRIPTION
**Pull request by vaishakp**
Add script to test the computation of observer waveform polarizations against lal.

This must be merged (in order) after #32, #35, and preferably before #34.

#25

### CODE REVIEW

File: .github/workflows/python-test.yml [hunk 1]
```diff
@@ -54,7 +54,7 @@ jobs:
             # ########################
             ## lal extras fetch method
             #git lfs pull
-            # wget https://git.ligo.org/lscsoft/lalsuite-extra/-/raw/master/data/lalsimulation/SEOBNRv4ROM_v2.0.hdf5 -O SEOBNRv4ROM_v2.0.hdf5
+            wget https://git.ligo.org/lscsoft/lalsuite-extra/-/raw/master/data/lalsimulation/SEOBNRv4ROM_v2.0.hdf5 -O SEOBNRv4ROM_v2.0.hdf5
```
Comment (prayush) [inline .github/workflows/python-test.yml:57]: Pls remind me, what are we using `SEOBNRv4ROM` for? Can we replace it with `SEOBNRv4` which is a bit slower but probably still faster than downloading this file.

File: test/test_gt_pol.py [hunk 2]
```diff
@@ -0,0 +1,509 @@
+""" Test the generation of observer polarizations
```
Comment (prayush) [inline test/test_gt_pol.py:1]: Please run `black` on this

File: test/test_gt_pol.py [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+    params['mass2'] = pnutils.mtotal_eta_to_mass1_mass2(params['mtotal'], params['eta'])[1]
+
+    # BH1 spins
+    params['spin1x'] = f.attrs['spin1x']
+    params['spin1y'] = f.attrs['spin1y']
+    params['spin1z'] = f.attrs['spin1z']
+
+
+    # BH2 spins
+
```
Comment (prayush) [inline test/test_gt_pol.py:196]: remove extra blank line

File: test/test_gt_pol.py [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+    # NR frame
+    s1 = [params['spin1x'], params['spin1y'], params['spin1z']]
+    s2 = [params['spin2x'], params['spin2y'], params['spin2z']]
+
+    # LAL frame
+    from nrcatalogtools.lvc import transform_spins_nr_to_lal
+    S1, S2 = transform_spins_nr_to_lal(s1, s2, params['nhat'], params['lnhat'])
+
+    from pycbc.waveform import get_td_waveform
+
```
Comment (prayush) [inline test/test_gt_pol.py:228]: Remove blank lines.

File: test/test_gt_pol.py [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+        params['coa_phase']=lal_coa_phase
+
+    # Transform spins
+
+    # NR frame
+    s1 = [params['spin1x'], params['spin1y'], params['spin1z']]
+    s2 = [params['spin2x'], params['spin2y'], params['spin2z']]
+
+    # LAL frame
+    from nrcatalogtools.lvc import transform_spins_nr_to_lal
```
Comment (prayush) [inline test/test_gt_pol.py:224]: Move all imports to the top of this file.

File: test/test_gt_pol.py [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+    if 'f_lower_at_1MSUN' not in list(f.attrs.keys()):
+        f.attrs['f_lower_at_1MSUN'] = f_lower_at_1MSUN
+        f.close()
+    else:
+        f.close()
+        
+    f = h5py.File(file, 'r')
+    # print(list(f.attrs.keys()))
+    # print(f.attrs.keys())
+    # print(f.keys())
```
Comment (prayush) [inline test/test_gt_pol.py:174]: Please move these prints behind a `debug` flag (or remove them).

File: test/test_gt_pol.py [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+
+    time_l = np.array(range( len(hp_l) ) )*delta_t
+    
+    # Recenter
+    mtime = time_l[np.argmax(hp_l**2 + hx_l**2)]
+    time_l -=mtime
+    # pyplot.figure()
+    # plt.plot(time_l, hp_l, color=[0,0.7071,1])
+    # plt.plot(time_l, hx_l, color=[0.1,0,0])
+    # plt.show()
```
Comment (prayush) [inline test/test_gt_pol.py:255]: Please move these plotting commands behind a `debug` flag (or remove them).

File: test/test_gt_pol.py [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+            against  waveformtools and lal. Tested are RMS errors, maximum deviation and mismatches.
+        """
+
+        inc_angles = [0.0015, 0.5, 1, 1.5, np.pi/2 - 0.001, np.pi/2, np.pi/2+0.0001, 2.5, 3, np.pi-0.0015] 
+
+        # inc_angles = np.arange(0, np.pi, 25)
+        coa_phases = np.linspace(0, 2*np.pi, 10)
+        # inc_angles = [np.pi/3]
+        # coa_phases = [np.pi/6]
+        # inc_angles = [0.5, 1, 1.5]
```
Comment (prayush) [inline test/test_gt_pol.py:352]: Please remove these commented out lines.

File: test/test_gt_pol.py [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+                    prec,
+                    f"The arrays (nrcatalogtools, lal) must equal atleast upto {prec} decimals",
+                )
+
+                
+
+                if test_wrt_wftools is True:
+            
+                    prec = 2
+                    self.assertAlmostEqual(
```
Comment (prayush) [inline test/test_gt_pol.py:463]: Run `black` on this file

File: test/test_gt_pol.py [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+                    np.array(wf_w3),
+                    prec,
+                    f"The arrays (nrcatalogtools, waveformtools) must equal atleast upto {prec} decimals",
+                    )
+        
+
+        max_mm = max(all_mm)
+
+        message(f"Max mismatch from this set is {max_mm}", message_verbosity=1)
+    if __name__ == "__main__":
```
Comment (prayush) [inline test/test_gt_pol.py:508]: Add a blank line right above

File: test/test_gt_pol.py [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+
+    # NR frame
+    s1 = [params['spin1x'], params['spin1y'], params['spin1z']]
+    s2 = [params['spin2x'], params['spin2y'], params['spin2z']]
+
+    # LAL frame
+    from nrcatalogtools.lvc import transform_spins_nr_to_lal
+    S1, S2 = transform_spins_nr_to_lal(s1, s2, params['nhat'], params['lnhat'])
+
+    from pycbc.waveform import get_td_waveform
```
Comment (prayush) [inline test/test_gt_pol.py:227]: Move all imports to the top of this file.

File: test/test_gt_pol.py [hunk 12]
```diff
@@ -0,0 +1,509 @@
...
+
+
+    print('Loading waveform through LAL')
```
Comment (prayush) [inline test/test_gt_pol.py:230]: -> 'Loading waveform through PyCBC'

**Review by prayush — CHANGES_REQUESTED** [review]

File: test/test_gt_pol.py [hunk 13]
```diff
@@ -0,0 +1,530 @@
...
+        coa_phase=coa_phase,
+        delta_t=delta_t,
+    )
+    hpc_pycbc = hpc  # mwf.to_pycbc(hpc)
+
+    # Minus sign to rotate by pi/2
+    hp_n, hx_n = hpc_pycbc.real(), hpc_pycbc.imag()
+
+    if nrcat_use_lal_conven is True:
+        hp_n, hx_n = hp_n, -hx_n
```
Comment (prayush) [inline test/test_gt_pol.py:132]: Would this be needed now that #36 has been merged?

File: .github/workflows/python-test.yml (hunk 1 above)
Comment (prayush) [reply .github/workflows/python-test.yml:57]: Also, shorten the path a few lines above (line 49):
```
export LAL_DATA_PATH="${PWD}/test_data/lalsuite/lalsuite-extra/data/lalsimulation"
```
to 
```
export LAL_DATA_PATH="${PWD}/lalsuite-extra/data/lalsimulation"
```

File: test/test_gt_pol.py (hunk 13 above)
Comment (vaishakp) [reply test/test_gt_pol.py:132]: No

File: .github/workflows/python-test.yml (hunk 1 above)
Comment (prayush) [reply .github/workflows/python-test.yml:57]: Finally, how is [this test](https://github.com/gwnrtools/nr-catalog-tools/actions/runs/5023480264/jobs/9008119973) failing even after you enabled the download of the required file?

File: test/test_gt_pol.py (hunk 13 above)
Comment (prayush) [reply test/test_gt_pol.py:132]: Excellent, in that case please remove this if-block, and if not needed, remove the flag `nrcat_use_lal_conven` entirely.

File: test/test_gt_pol.py [hunk 14]
```diff
@@ -0,0 +1,494 @@
...
+    hpc = mwf1.get_td_waveform(
+        total_mass=total_mass,
+        distance=distance,
+        inclination=inclination,
+        coa_phase=coa_phase,
+        delta_t=delta_t,
+    )
+    hpc_pycbc = hpc
+
+    # Minus sign to rotate by pi/2
```
Comment (prayush) [inline test/test_gt_pol.py:119]: what minus sign?

File: test/test_gt_pol.py [hunk 15]
```diff
@@ -0,0 +1,494 @@
...
+
+message(f"Simulation {sim_name}", message_verbosity=2)
+
+wftools_use_lal_conven = True
+lal_use_coa_phase_as_phi_ref = False
+test_wrt_wftools = False
+
+######################################
+# Waveform comparison function
+######################################
```
Comment (prayush) [inline test/test_gt_pol.py:73]: delete these :)

**Review by prayush — CHANGES_REQUESTED** [review]
as always, run `black` please!

**Pull request merged**